cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -1204,34 +1139,36 @@ if(ARROW_JEMALLOC)
       "${CMAKE_CURRENT_BINARY_DIR}/jemalloc_ep-prefix/src/jemalloc_ep/dist/")
   set(JEMALLOC_STATIC_LIB
       "${JEMALLOC_PREFIX}/lib/libjemalloc_pic${CMAKE_STATIC_LIBRARY_SUFFIX}")
-  # We need to disable TLS or otherwise C++ exceptions won't work anymore.
-  externalproject_add(
-    jemalloc_ep
-    URL
-    ${CMAKE_CURRENT_SOURCE_DIR}/thirdparty/jemalloc/${JEMALLOC_VERSION}.tar.gz
-    PATCH_COMMAND
-    touch
-    doc/jemalloc.3
-    doc/jemalloc.html
-    CONFIGURE_COMMAND
-    ./autogen.sh
-    "AR=${CMAKE_AR}"
-    "CC=${CMAKE_C_COMPILER}"
-    "--prefix=${JEMALLOC_PREFIX}"
-    "--with-jemalloc-prefix=je_arrow_"
-    "--with-private-namespace=je_arrow_private_"
-    "--disable-tls"
-    ${EP_LOG_OPTIONS}
-    BUILD_IN_SOURCE
-    1
-    BUILD_COMMAND
-    ${MAKE}
-    ${MAKE_BUILD_ARGS}
-    BUILD_BYPRODUCTS
-    "${JEMALLOC_STATIC_LIB}"
-    INSTALL_COMMAND
-    ${MAKE}
-    install)
+  externalproject_add(jemalloc_ep
+                      URL
+                      ${JEMALLOC_SOURCE_URL}
+                      PATCH_COMMAND
+                      touch
+                      doc/jemalloc.3
+                      doc/jemalloc.html
+                      CONFIGURE_COMMAND
+                      ./autogen.sh
+                      "AR=${CMAKE_AR}"
+                      "CC=${CMAKE_C_COMPILER}"
+                      "--prefix=${JEMALLOC_PREFIX}"
+                      "--with-jemalloc-prefix=je_arrow_"
+                      "--with-private-namespace=je_arrow_private_"
+                      "--without-export"
+                      # Don't override operator new()
```

Comment:
Do we have an explanation why this is problematic or even an upstream issue?

Reply:
My guess is that, when `libarrow.so` is used from external code, the external code will not get the `operator new` override. So if e.g. a `Status` is allocated inside `libarrow.so` and deleted by that external code, or vice-versa, you get one allocator trying to deallocate memory allocated by another allocator.

More generally, since Arrow is a library, it shouldn't muck with the default C and C++ allocators.